feat(SideNav): implement 3rd level navigation

IgnacioBecerra opened the pull request:
Closes https://github.com/carbon-design-system/carbon/issues/18466
Closes #415 

This PR implements the 3rd level navigation for the SideNav component, as well as the Panel navType feature where the submenu icon will link to its first link when the sidenav is in a collapsed state. 

The treeview state is enabled when the prop `isTreeview` is added to the `SideNav`. If there is a nested sub-menu, this feature will be enabled automatically.

#### Changelog

**New**

- TreeView navigation for UIShell
- added SideNavMenu nesting feature
- added keyboard arrow navigation
- fixed styles for Carbon Labs UIShell
- sub-menu links to first link in a collapsed SideNav

#### Testing / Reviewing
Ran `npm test`, also test using keyboard navigation

alisonjoseph commented:
One thing I wanted to make sure was clear, we only want to keyboard nav to switch to treeview if it has a 3rd level, otherwise it should function as it does today with tabs.

alisonjoseph commented:
Copying over @oliviaflory comments  from https://github.com/carbon-design-system/carbon-labs/issues/419 related to the 3rd level nav here


### Styles


2. Sub menu level 1: Both icons should change to $icon-primary on hover

![Image](https://github.com/user-attachments/assets/696360eb-23b5-4f69-b682-8e82ee604465)

3. Sub menu level 2: chevron should change to $icon-primary on hover

![Image](https://github.com/user-attachments/assets/a72793b1-1ff5-43c4-962c-127031dd3409)

4. Icon of parent menu should be $icon-primary when sub level link is selected

![Image](https://github.com/user-attachments/assets/f9023216-112e-4401-8a9e-8cc07f3b2e31)


## Side nav

### Keyboard

1. Cannot access the expand button via keyboard when the side nav is collapsed

2. Cannot access the collapse button via keyboard when the side nav is expanded

### Styles

1. Icon in expanded state should change to $icon-primary on hover

![Image](https://github.com/user-attachments/assets/c1c12331-64ef-44c2-b68f-9cb8816c7dfd)

2. Icon in collapsed state should change to $icon-primary on hover

![Image](https://github.com/user-attachments/assets/f810a88f-b3ef-408d-9347-33270bba3033)

IgnacioBecerra commented:
PR is ready for review! All the new features we published are included, with fixes and improvements. Also fixed some style issues.

cc: @oliviaflory - please review when you have time next week! I addressed your comments from that bigger issue.

oliviaflory commented:
@IgnacioBecerra the styles are looking good, thanks for fixing those!

This is a bit nit-picky but anything that has additional items should probably be called `sub-menu`, we should reserve `link` for a single click target (no nested items). Seeing this in the [Default With Tree View story](https://deploy-preview-410--carbon-labs-react.netlify.app/?path=/story/components-uishell--default-with-tree-view) and some of the side nav stories, but maybe side nav isn't part of this PR
<img width="302" alt="Screenshot 2025-02-10 at 10 40 31 AM" src="https://github.com/user-attachments/assets/84030cf0-4e02-49e5-ab63-ec47b74e5d9a" />

alisonjoseph commented:
Couple functional things I'm seeeing. I can't tab into SubMenuItems anymore in the SideNav without tree

https://github.com/user-attachments/assets/77e4efac-2180-4274-94ae-8761fe13a555


I'm also seeing an extra keyboard tab stop in the DefaultWithTreeView story. After hitting enter on the hamburger to open the menu, I have to tab twice to get to the first link, this doesn't happen with the non-tree sidenav.

alisonjoseph reviewed (changes requested):
Tabbing is working and the double tab issue is also gone now. I ran the a11y checker and there are several errors to look through. 

<img width="748" alt="Screenshot 2025-02-11 at 7 55 57 AM" src="https://github.com/user-attachments/assets/23d8435b-c641-4d29-bf40-1d35fad2981a" />



- I believe you will want to move role=tree to the ul and off of nav
- add an aria-label to the ul
- and for the tabbable error, I think its complaining about the buttons for the submenu all having tabindex="-1"

IgnacioBecerra commented:
@alisonjoseph Hm, adding another tabindex to the `<li role="treeitem">` would make it so the focus gets "doubled", since the real focus should be within the `<a>`. For the Treeview component, all the focus styles are in the `<li>`, so that's where it differs. Thoughts about this?

alisonjoseph commented:
@IgnacioBecerra I'm now wondering if the role=treeitem should actually be on the a tag/link instead of the li? That appears to be how they are doing it in the example you can open in codepen here https://www.w3.org/WAI/ARIA/apg/patterns/treeview/examples/treeview-navigation/

alisonjoseph commented:
Couple other things, I'd like to add at least some minimal docs to the overview page for this before merging. 

- Update default story to use type treeview, can then remove the default with tree story
- Update stackblitz example to use type treeview
- Update overview page
  - Update main code sample
  - Add new section for treeview with a note around when to use type treeview (when more than 2 levels) with a code sample(can probably just use the code from sidenav with third level story).
  - Should we re-name sidenav with third level to SideNav TreeView?

alisonjoseph commented:
@IgnacioBecerra I'm fine with keeping the story name you got from Olivia 😄

emyarod commented:
> @IgnacioBecerra I'm now wondering if the role=treeitem should actually be on the a tag/link instead of the li? That appears to be how they are doing it in the example you can open in codepen here https://www.w3.org/WAI/ARIA/apg/patterns/treeview/examples/treeview-navigation/
> 
> @emyarod thoughts on this. The navigation as a tree is a little different than the general treeview component because it has links.

yeah I believe the treeitem role goes on the link like in the w3 example

alisonjoseph commented on packages/react/src/components/UIShell/components/SideNavMenuItem.tsx:
```diff
@@ -87,16 +89,16 @@ export const SideNavMenuItem = React.forwardRef<
       linkRef.current.style.paddingLeft = `${calcLinkOffset()}rem`;
     }
   }, []);
+  console.log(navType);
```
```suggestion
```

alisonjoseph commented on packages/react/src/components/UIShell/components/styles/_side-nav.scss:
```diff
@@ -48,12 +48,22 @@ div:has(.#{$prefix}--header)
 
-//----------------------------------------------------------------------------
+.#{$prefix}--side-nav__link,
+.#{$prefix}--side-nav__submenu {
+  &:hover {
+    .#{$prefix}--side-nav__icon > svg,
+    .#{$prefix}--side-nav__submenu-chevron > svg {
+      fill: $icon-primary;
+    }
+  }
+}
```
```suggestion
.#{$prefix}--side-nav__submenu:hover {
  .#{$prefix}--side-nav__icon > svg,
  .#{$prefix}--side-nav__submenu-chevron > svg {
    fill: $icon-primary;
  }
}
```

super nitpicky, but we try to avoid nesting more than once in general in scss files in Carbon.

alisonjoseph commented on packages/react/src/components/UIShell/components/styles/_side-nav.scss:
```diff
@@ -63,6 +73,14 @@ div:has(.#{$prefix}--header)
     margin-inline-end: $spacing-05;
   }
 
+  .#{$prefix}--side-nav__item.#{$prefix}--side-nav__link {
+    &:hover {
+      .#{$prefix}--side-nav__icon > svg {
+        fill: $icon-primary;
+      }
+    }
+  }
```
Same thing here

emyarod commented:
does the user need to set the navType to treeview when there are nested menus or can that be automatically set?

emyarod commented on packages/react/src/components/UIShell/components/SideNav.tsx:
```diff
@@ -34,6 +34,7 @@ export enum SIDE_NAV_TYPE {
   DEFAULT = 'default',
   RAIL = 'rail',
   PANEL = 'panel',
+  TREEVIEW = 'treeview',
```
I think we briefly discussed this during a dev sync but does the user need to set the navType to treeview when there are nested menus or can that be automatically set?

IgnacioBecerra replied:
For the moment it needs to be automatically set, but we'd have to discuss about the requirements for the automatic setting. Mainly, would the existence of a nested menu mean that the whatever `navType` will be overridden? 

cc: @alisonjoseph @oliviaflory

emyarod replied:
yeah I think having it tied to the role makes sense and we can revisit this in a future PR :+1:

alisonjoseph commented:
Just noticed one other thing when I ran the a11y checker on the default SideNav story. Looks like the li still have role="treeitem" and its causing an error.

<img width="705" alt="Screenshot 2025-02-13 at 8 08 25 AM" src="https://github.com/user-attachments/assets/1e928bb9-4197-454c-938a-fd4729b8f52f" />

alisonjoseph commented on examples/react/UIShell/src/App.tsx:
```diff
@@ -194,8 +194,8 @@ function App() {
               <SideNav
-                navType={SIDE_NAV_TYPE.PANEL}
                 isChildOfHeader={false}
+                isTreeview={true}
```
```suggestion
```
I don't think we want this for this secondary nav?

IgnacioBecerra replied:
woops, yeah i'll remove

alisonjoseph commented on packages/react/src/components/UIShell/__stories__/UIShell.mdx:
```diff
@@ -207,6 +202,62 @@ function App() {
+`isTreeview` prop. Note that this feature should only be used if there are two
+levels of content within the component.
+
+Also mote that it will be automatically enabled if there are more than two levels of
```
```suggestion
Also note that it will be automatically enabled if there are more than two levels of
```

alisonjoseph commented on packages/react/src/components/UIShell/components/SideNavItems.tsx:
```diff
@@ -7,11 +7,13 @@
+import { SideNavContext } from './SideNav';
 
 export interface SideNavItemsProps {
+  accessibilityLabel: object;
```
Can we add comments explaining what this is for?

IgnacioBecerra commented:
Expected functionality for #415:

https://github.com/user-attachments/assets/10d45070-bfe6-4ec0-af62-ce27bda3844a

Saves menu expanded state so it's back to where you left it if you collapse sidenav and then reopen.

alisonjoseph reviewed (approved):
LGTM!

IgnacioBecerra commented:
@oliviaflory  thanks for the review! We can follow up on those issues in a later PR as they're not related to the tree view features but the Panel view instead. 

For the first issue, I think it would make sense for the previous tab to go to the hamburger menu, but the panel technically part of the page body instead of the menu bar, so it also makes sense for it to go to the last element in the menu...

As for the second issue, I'd say it might be because there's no content, but this one does make sense for the focus to go inside the right panel like in regular carbon.

Maybe we should go over that first issue with Michael Gower? 

Either way, seems like tree view is good design wise, thank you!

cc: @alisonjoseph

oliviaflory commented:
@IgnacioBecerra oh I see for the first question the Carbon core story behaves the same way, so that makes sense. I didn't realize the main menu panel was part of the page body, thanks for explaining!

oliviaflory reviewed (approved):
LGTM 🎉
